dimod/sampleset.py

```diff
@@ -997,6 +1002,90 @@ def aggregate(self):
+
+            sort_labels (bool, optional, default=True):
+                If true, returned :attr:`.SampleSet.variables` will be in
+                sorted-order. Note that mixed types are not sortable in which
+                case the given order will be maintained.
+
+        Returns:
+            :obj:`.SampleSet`: A new sample set with the samples appended. If
+            `samples_like` is another sample set then the returned sample set
+            has the sum of the energies.
+
+        Examples:
+
+            >>> sampleset = dimod.SampleSet.from_samples([{'a': -1, 'b': +1},
+            ...                                           {'a': +1, 'b': +1}],
+            ...                                          dimod.SPIN,
+            ...                                          energy=[-1.0, 1.0])
+            >>> new = sampleset.append_variables({'c': -1})
+            >>> print(new)
+               a  b  c energy num_oc.
+            0 -1 +1 -1   -1.0       1
+            1 +1 +1 -1    1.0       1
+            ['SPIN', 2 rows, 2 samples, 3 variables]
+
+            Append another sampleset to the original above.
+
+            >>> another = dimod.SampleSet.from_samples([{'c': -1, 'd': +1},
+            ...                                         {'c': +1, 'd': +1}],
+            ...                                        dimod.SPIN,
+            ...                                        energy=[-2.0, 1.0])
+            >>> new = sampleset.append_variables(another)
+            >>> print(new)
+               a  b  c  d energy num_oc.
+            0 -1 +1 -1 +1   -3.0       1
+            1 +1 +1 +1 +1    2.0       1
+            ['SPIN', 2 rows, 2 samples, 4 variables]
+
+        """
+        samples, labels = as_samples(samples_like)
+
+        num_samples = len(self)
+
+        # we don't handle multiple values
+        if samples.shape[0] == num_samples:
+            # we don't need to do anything, it's already the correct shape
+            pass
+        elif samples.shape[0] == 1 and num_samples:
+            samples = np.repeat(samples, num_samples, axis=0)
+        else:
+            msg = ("mismatched shape. The samples to append should either be "
+                   "a single sample or should match the length of the sample "
+                   "set. Empty sample sets cannot be appended to.")
+            raise ValueError(msg)
+
+        # append requires the new variables to be unique
+        variables = self.variables
+        if any(v in variables for v in labels):
+            msg = "Appended samples cannot contain variables in sample set"
+            raise ValueError(msg)
+
+        new_variables = list(variables) + labels
+        new_samples = np.hstack((self.record.sample, samples))
+
+        new = type(self).from_samples((new_samples, new_variables),
+                                      self.vartype,
+                                      info=dict(self.info),  # make a copy
```

Comment:
Do you want to actually `deepcopy` here?

Reply:
Good question. In my mind that's on the user, but I suppose there is no harm in deepcopying

Reply:
Since we already treat `SampleSet` as an immutable object, deepcopying makes more sense here (despite the performance penalty). Just consider user has mutable values in `info` dict... Now after the append, they would get a fresh copy of `SampleSet`, but the values in their `info` dicts would point to the same container(s).

Reply:
Also fixed this in other methods.